autoprecompiles/src/optimizer_documentation.md

```diff
@@ -0,0 +1,221 @@
+# The Autoprecompiles Optimizer
+
+## Terminology
+
+### Field Elements
+
+Throughout this document, we will be working in a finite field of prime order `p`. Often, we use operators or concepts that are only defined
+in the integers. In this case, we use the natural number representation
+of the field element, i.e. the unique integer `x` such that `0 <= x < p`
+where the field operations are defined as `x + y = (x + y) mod p`
+and `x * y = (x * y) mod p` for field elements `x` and `y`.
+
+This way, we can also make statements about a field element being
+_less than_ another field element, even if this would not make sense
+inside the finite field. Sometimes, field elements are also interpreted
+as signed integers instead of natural numbers, but this will be clarified.
+
+### Constraint System
+
+The optimizer is operating on an abstraction of a chip we call
+_Constraint System_, which consists of a set of _Algebraic Constraints_
+and _Bus Interactions_.
+
+### Algebraic Constraint
+
+An _Algebraic Constraint_ consists of an _Algebraic Expression_, i.e.
+an expression involving the operators `+` and `*` on _Variables_ and
+_Constants_ from the implied finite field. The idea is that the constraint
+forces the expression to be zero and thus we write it as an equation
+`<expr> = 0` (but also `<lhs> = <rhs>` if more convenient, by which
+we mean `<rhs> - <lhs> = 0`).
+
+An Algebraic Constraint is _satisfied_ by an assignment of the variables
+if it evaluates to zero under this assignment.
+
+Example: `x * (x - 1) = 0` is an algebraic constraint that forces
+the variable `x` to be either zero or one, meaning that an
+assignment only satisfies this constraint if it has `x = 0` or `x = 1`.
+
+### Range Constraint
+
+The task of the optimizer is hugely simplified by the concept of
+_Range Constraints_. In an abstract way, a _Range Constraint_ is
+the set of possible values for a specific algebraic expression.
+The solver can derive new Range Constraints from Algebraic Constraints
```

Comment:
This is the first time `solver` is used, it'd be good to refer to something or define it.